### PR TITLE
Support for EIP-2612: EIP712-signed token approvals

### PR DESCRIPTION
**Pull request by lucasvo**
This implements https://github.com/ethereum/EIPs/pull/2612 mostly inspired by uniswap's ERC20 code.

### CODE REVIEW

File: src/erc20.sol [hunk 1]
```diff
@@ -97,4 +116,20 @@ contract ERC20 {
     function move(address src, address dst, uint wad) external {
         transferFrom(src, dst, wad);
     }
+
+    // --- Approve by signature ---
+    function permit(address owner, address spender, uint value, uint deadline, uint8 v, bytes32 r, bytes32 s) external {
```
Comment (xmxanuel) [inline src/erc20.sol:121]: The uniswap implementation is used as reference here, correct?
https://github.com/Uniswap/uniswap-v2-core/blob/master/contracts/UniswapV2ERC20.sol#L81

File: src/erc20.sol [hunk 2]
```diff
@@ -97,4 +116,20 @@ contract ERC20 {
...
+        require(deadline >= block.timestamp, 'cent/past-deadline');
+        bytes32 digest = keccak256(
+            abi.encodePacked(
+                '\x19\x01',
+                DOMAIN_SEPARATOR,
+                keccak256(abi.encode(PERMIT_TYPEHASH, owner, spender, value, nonces[owner]++, deadline))
+            )
+        );
+        address recoveredAddress = ecrecover(digest, v, r, s);
+        require(recoveredAddress != address(0) && recoveredAddress == owner, 'cent-erc20/invalid-sig');
```
Comment (xmxanuel) [inline src/erc20.sol:131]: I am curious why a check `recoveredAddress != address(0)` is required. This is the only check i don't understand. Need to look a bit more into EIP-2612.

Comment (xmxanuel) [reply src/erc20.sol:131]: Ah I see.  `address(0)` is the error return from `ecrecover`

https://solidity.readthedocs.io/en/latest/units-and-global-variables.html#mathematical-and-cryptographic-functions

**Review by xmxanuel — APPROVED** [review]
LGTM.

PR is adopting the uniswap `permit` implementation.

File: src/erc20.sol [hunk 3]
```diff
@@ -29,9 +29,13 @@ contract ERC20 {
     string  public version;
     uint256 public totalSupply;
 
+    bytes32 public DOMAIN_SEPARATOR;
```
Comment (MrChico) [inline src/erc20.sol:32]: can be made immutable

**Pull request merged**